shopinvader_validation_token/models/shopinvader_security_token.py

```diff
@@ -0,0 +1,162 @@
+        Should be used in a cron.
+        :return:
+        """
+        now = fields.Datetime.now()
+        domain = [
+            "|",
+            ("active", "=", False),
+            ("expiration_datetime", "<", now),
+        ]
+        shopinvader_tokens = self.sudo().search(domain)
+        if shopinvader_tokens:
+            shopinvader_tokens.unlink()
+
+    @api.model
+    def _service_notification_map(self):
+        """
+
+        :return: dict
+        """
+        return {
+            "shopinvader.customer.service": "partner_token_customer",
+        }
+
+    @api.model
+    def _should_trigger_security_token(self, backend, email, service_name):
+        """
+        Check if the notification partner_token_* is enabled
+        on the related backend.
+        As we have the backend, email and the service who do this check,
+        we can precisely check if the token is required or not.
+        :param backend: shopinvader.backend recordset
+        :param email: str
+        :param service_name: str
+        :return: bool
+        """
+        # We have to force the backend when the current shopinvader.partner
+        # doesn't exist yet.
+        backend.ensure_one()
+        should_trigger = False
+        if service_name and backend:
+            notif_type = self._service_notification_map().get(service_name, "")
+            if notif_type:
+                should_trigger = any(
+                    n.notification_type == notif_type
+                    for n in backend.notification_ids
+                )
+        return should_trigger
+
+    @api.multi
+    def _trigger_token_notification(self):
+        """
+        Generate the shopinvader.security.token and trigger the notification
+        on it.
+        :return:
+        """
+        self.ensure_one()
+        self.shopinvader_backend_id._send_notification(
+            self.notification_type, self
+        )
+
+    @api.model
+    def _get_new_token(self, backend, email):
+        """
+        Generate a new token/code
+        :param backend: shopinvader.backend recordset
+        :param email: str
+        :return: str
+        """
+        return u"{:06d}".format(randint(0, 999999))
+
+    @api.model
+    def _generate_token(self, backend, email, service_name):
+        """
+        Generate the shopinvader.security.token
+        :return: shopinvader.security.token
+        """
+        token = self._get_new_token(backend, email)
+        now = fields.Datetime.from_string(fields.Datetime.now())
+        expiration = now + relativedelta(minutes=backend.token_validity)
+        expiration_datetime = fields.Datetime.to_string(expiration)
+        notif_type = self._service_notification_map().get(service_name)
+        values = {
+            "shopinvader_backend_id": backend.id,
+            "email": email,
+            "token": token,
+            "expiration_datetime": expiration_datetime,
+            "notification_type": notif_type,
```

Comment:
instead of binding this to `notification_type` wouldn't be better to have `token_type`?
The notification type can be `service_name:token_type`.

Reply:
We should probably have an independent token configuration instead.
With a new model, with a python snippet that allows to hooks this machinery via configuration in any service.
What is your main goal? Validate emails? Validate important changes?

Reply:
what about this?